source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +117,185 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
 
+  // Serializes the generated response to the parsed query from the client. If there is a
+  // parsing error or the incoming query is invalid, we will still generate a valid DNS response
+  message_parser_.buildResponseBuffer(query_context, response);
   Network::UdpSendData response_data{query_context->local_->ip(), *(query_context->peer_),
                                      response};
   listener_.send(response_data);
 }
 
+DnsLookupResponseCode DnsFilter::getResponseForQuery(DnsQueryContextPtr& context) {
+  /* It appears to be a rare case where we would have more than one query in a single request.
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
+    // always attempt to resolve with the configured domains
+    if (isKnownDomain(query->name_) || !config_->forwardQueries()) {
+      // Determine whether the name is a cluster. Move on to the next query if successful
+      if (resolveViaClusters(context, *query)) {
+        continue;
+      }
+
+      // Determine whether we an answer this query with the static configuration
+      if (resolveViaConfiguredHosts(context, *query)) {
+        continue;
+      }
+    }
+    // TODO(abaptiste): resolve the query externally
+  }
+
+  if (context->answers_.empty()) {
+    return DnsLookupResponseCode::Failure;
+  }
+  return DnsLookupResponseCode::Success;
+}
+
+uint32_t DnsFilter::getDomainTTL(const absl::string_view domain) {
+  const auto& domain_ttl_config = config_->domainTtl();
+  const auto& iter = domain_ttl_config.find(domain);
+
+  uint32_t ttl;
+  if (iter == domain_ttl_config.end()) {
+    ttl = static_cast<uint32_t>(DEFAULT_RESOLVER_TTL.count());
+  } else {
+    ttl = static_cast<uint32_t>(iter->second.count());
+  }
+
+  return ttl;
+}
+
+bool DnsFilter::isKnownDomain(const absl::string_view domain_name) {
+  const auto& known_suffixes = config_->knownSuffixes();
+
+  // If we don't have a list of whitelisted domain suffixes, we will resolve the name with an
+  // external DNS server
+  if (known_suffixes.empty()) {
+    ENVOY_LOG(debug, "Known domains list is empty");
+    return false;
+  }
+
+  // TODO(abaptiste): Use a trie to find a match instead of iterating through the list
+  for (auto& suffix : known_suffixes) {
+    if (suffix->match(domain_name)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+const DnsEndpointConfig* DnsFilter::getEndpointConfigForDomain(const absl::string_view domain) {
+  const auto& domains = config_->domains();
+
+  // TODO: If we have a large ( > 100) domain list, use a binary search.
```

Comment:
Why not just use an absl::flat_hash_map now?

Reply:
It is :)  I will remove this comment.